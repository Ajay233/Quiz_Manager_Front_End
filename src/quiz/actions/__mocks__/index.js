export const setQuiz = jest.fn()
export const createQuiz = jest.fn( () => { return {type: "SET_QUIZ", payload:""} })
export const updateQuiz = jest.fn( () => { return {type: "SET_QUIZ", payload:""} })
export const updateQuizStatus = jest.fn()
