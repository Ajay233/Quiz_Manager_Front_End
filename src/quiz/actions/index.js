import { get, getUsingParams } from '../../axiosRequests/requests';

export const setQuizes = (quizes) => {
  return {
    type: "SET_QUIZES",
    payload: quizes
  }
}

export const getAllQuizes = (endpoint, data, token) => {
  return (dispatch) => {
    get(endpoint, data, token).then((response) => {
      console.log(response)
      dispatch({
        type: "GET_ALL_QUIZES",
        payload: response.data
      })
    }).catch((error) => {
      console.log(error.response)
    });
  }
}

export const getQuizSearchResults = (endpoint, data, token, setNotification, errorMsg) => {
  return (dispatch) => {
    getUsingParams(endpoint, data, token).then((response) => {
      dispatch({
        type: "SET_QUIZ_SEARCH_RESULTS",
        payload: response.data
      })
    }).catch((error) => {
      console.log(error.response);
      setNotification(errorMsg, "error", true);
    })
  }
}
