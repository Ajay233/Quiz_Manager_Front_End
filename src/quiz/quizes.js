import React from 'react'
import Quiz from './quiz'

class Quizes extends React.Component {



  render(){
    return(
      <div>

      </div>
    );
  }
}

export default Quizes
