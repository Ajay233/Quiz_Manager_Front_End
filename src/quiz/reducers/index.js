import { quiz } from '../factory/quizFactory'

export const setQuizReducer = (state=quiz, action) => {
  switch (action.type) {
    case "SET_QUIZ": return action.payload;
    default: return state;
  }
}
