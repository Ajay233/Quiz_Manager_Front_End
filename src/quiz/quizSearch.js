import React from 'react';
import { connect } from 'react-redux'
import Quizes from './quizes'
import QuizSearchByName from '../forms/findQuizByName'
import Notification from '../notifications/notifications'
import { setNotification } from '../notifications/actions'
import { setQuizes, getQuizSearchResults } from './actions'
import "../stylesheets/quizSearch.css"


class QuizSearch extends React.Component {

  renderResultsTitle = (quizes) => {
    return quizes.length === 0 ? null : <div id="resultHeader">{`${quizes.length} Results`}</div>;
  }

  renderResultHeadings = (quizes) => {
    return quizes.length === 0 ? null : this.resultHeadings()
  }

  resultHeadings = () => {
    return(
      <div id="resultsHeadings">
        <div id="nameHeader">Quiz Name</div>
        <div id="descriptionHeader">Description</div>
        <div id="optionsHeader">Options</div>
      </div>
    )
  }

  render(){
    const { quizes, userData, setNotification, getQuizSearchResults } = this.props
    return(
      <div id="quizSearch">
        <Notification />
        <div id="quizSearchTitle">Quiz Search</div>
        <div id="quizSearchByName">
          <QuizSearchByName jwt={userData.jwt} setNotification={setNotification} getQuizSearchResults={getQuizSearchResults} />
        </div>
        <div id="quizSearchByCategory">

        </div>
        <div>
          {this.renderResultsTitle(quizes)}
          {this.renderResultHeadings(quizes)}
          <Quizes quizes={quizes} permission={userData.permission} />
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    userData: state.userData,
    notificationData: state.notificationData,
    quizes: state.quizes
  }
}

export default connect(mapStateToProps, { setQuizes, getQuizSearchResults, setNotification })(QuizSearch)
