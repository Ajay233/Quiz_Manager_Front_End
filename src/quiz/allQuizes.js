import React from 'react'
import { connect } from 'react-redux'
import Notification from '../notifications/notifications'
import QuizResults from '../quizSearch/quizResults'
import Modal from '../modal/modal'

import { getAllQuizes, deleteQuizFromCategory, clearQuizes } from '../quizSearch/actions'
import { hideModal } from '../modal/actions'
import { setNotification } from '../notifications/actions'
import { getQuestions } from '../question/actions'
import { del } from '../axiosRequests/requests'
import { sessionExpired } from '../utils/session'
import history from '../history'

import "../stylesheets/allQuizzes.css"

export class AllQuizes extends React.Component {

  // componentWillUnmount(){
  //   this.props.clearQuizes();
  // }

  renderModal = () => {
    const { hideModal } = this.props
    const { showModal, showModal2 } = this.props.modalState
    if(showModal){
      return(
        <Modal
          show={showModal}
          title={"Delete Quiz"}
          message={"You are about to delete a quiz which will also delete any questions and answers associated with it.  Do you want to continue?"}
          onDelete={this.handleDelete}
          onCancel={hideModal}
        />
      );
    } else if(showModal2){
      return(
        <Modal
          type={"start"}
          show={showModal2}
          title={"Start Quiz"}
          message={"You are about to start a quiz, would you like to continue?"}
          onStart={this.handleStartQuiz}
          onCancel={hideModal}
        />
      );
    } else {
      return null
    }
  }

  renderCategoryIcon = (category) => {
    switch (category) {
      case "Comics": return <div><img className="categoryIcon" src={require("../public/icons/comics.png")} alt=""/></div>
      case "TV": return <div><img className="categoryIcon" src={require("../public/icons/tv.png")} alt=""/></div>
      case "Gaming": return <div><img className="categoryIcon" src={require("../public/icons/games.png")} alt=""/></div>
      case "Films": return <div><img className="categoryIcon" src={require("../public/icons/films2.png")} alt=""/></div>
      case "Sports": return <div><img className="categoryIcon" src={require("../public/icons/sport.png")} alt=""/></div>
      case "Music": return <div><img className="categoryIcon" src={require("../public/icons/music.png")} alt=""/></div>
      case "History": return <div><img className="categoryIcon" src={require("../public/icons/museumV2.png")} alt=""/></div>
      case "Science": return <div><img className="categoryIcon" src={require("../public/icons/science.png")} alt=""/></div>
      case "Technology": return <div><img className="categoryIcon" src={require("../public/icons/technology.png")} alt=""/></div>
      case "Art": return <div><img className="categoryIcon" src={require("../public/icons/art.png")} alt=""/></div>
      case "Literature": return <div><img className="categoryIcon" src={require("../public/icons/books.png")} alt=""/></div>
      case "General Knowledge": return <div><img className="categoryIcon" src={require("../public/icons/generalKnowledge.png")} alt=""/></div>
      case "Food and Drink": return <div><img className="categoryIcon" src={require("../public/icons/Food-icon.png")} alt=""/></div>
      case "Geography": return <div><img className="categoryIcon" src={require("../public/icons/globe-icon.png")} alt=""/></div>
      default: return null
    }
  }

  renderQuizResults = (filteredQuizList, quizes, listItem, userData, setNotification) => {
    if(filteredQuizList.length > 0){
      return(
        <React.Fragment>
          {this.renderResultHeadings(filteredQuizList)}
          <QuizResults
            key={"Quiz" + quizes.indexOf(listItem)}
            quizes={filteredQuizList}
            permission={userData.permission}
            jwt={userData.jwt}
            setNotification={setNotification}
          />
        </React.Fragment>
      );
    } else {
      return <div id="noQuizzesMsg">There are currenty no quizzes for this category</div>
    }
  }

  // add contitional css alternate between left and right align classes
  // Will need a way to turn this on and off, possibly by using history.location.pathname
  // so this is only applied on the all quizes view
  renderCategories = () => {
    const { quizes, userData, setNotification } = this.props
    let quizesList = quizes.map(listItem => {
      let filteredQuizList = userData.permission === "USER" ? listItem.quizList.filter(quiz => quiz.status === "READY") : listItem.quizList;
      return(
        <div key={"QuizCategory" + quizes.indexOf(listItem)} className="quizCategorySection">
          <div className="categoryHeader">
            <div className="categoryTitle">{listItem.category}</div>
            <div>{this.renderCategoryIcon(listItem.category)}</div>
          </div>
          {this.renderQuizResults(filteredQuizList, quizes, listItem, userData, setNotification)}
        </div>
      );
    })
    return quizesList;
  }

  renderResultHeadings = (quizes) => {
    return quizes.length === 0 ? null : this.resultHeadings()
  }

  resultHeadings = () => {
    const { permission } = this.props.userData
    return(
      <div id="resultsHeadings" className="resultHeadings">
        <div className={permission !== "USER" ? "nameHeader" : "nameHeaderExpanded"}>Quiz Name</div>
        <div className="descriptionHeader">Description</div>
        { permission !== "USER" ? <div className="statusHeader">Status</div> : null }
      </div>
    )
  }

  handleDelete = () => {
    const { quiz, userData, deleteQuizFromCategory, setNotification, hideModal } = this.props;
    const config = {
      data: quiz
    }
    del("quiz/delete", config, userData.jwt).then((response) => {
      hideModal()
      deleteQuizFromCategory(quiz)
      setNotification("Quiz deleted", "success", true)
    }).catch((error) => {
      console.log(error.response)
      if(error.response.status === 403){
        sessionExpired(this.props.dispatch);
      } else {
        hideModal()
        setNotification("Error - Unable to delete this quiz", "error", true)
      }
    })
  }

  handleStartQuiz = () => {
    const { userData, quiz, getQuestions, hideModal } = this.props
    const param = { quizId: quiz.id }
    getQuestions("question/findByQuizId", param, userData.jwt, true)
    hideModal()
    history.push("/startQuiz")
  }

  renderComponent = () => {
    const { quizes } = this.props;
    if(quizes === null || quizes.length === 0 ){
      //This will need to be replaced with a component that renders an animated loading modal
      //Reuse modal
      return(
        <div>Loading...</div>
      );
    } else {
      return(
        <React.Fragment>
          <Notification />
          <div id="allQuizzesTitle">Browse all quizzes by Category</div>
          {this.renderModal()}
          {this.renderCategories()}
        </React.Fragment>
      );
    }
  }

  render(){
    return(
      <div className="componentContainer">
        {this.renderComponent()}
      </div>
    );
  }
}

export const mapStateToProps = (state) => {
  return {
    userData: state.userData,
    quizes: state.quizes,
    quiz: state.quiz,
    modalState: state.modalState
  }
}

export default connect(mapStateToProps,
  { getAllQuizes,
    deleteQuizFromCategory,
    getQuestions,
    hideModal,
    setNotification,
    clearQuizes
  })(AllQuizes)
