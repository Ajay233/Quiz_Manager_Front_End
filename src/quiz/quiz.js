import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import Notification from '../notifications/notifications'
import Questions from '../question/questions'
import UpdateQuizForm from '../forms/updateQuiz'
import Modal from '../modal/modal'
import { getQuestions, deleteQuestion } from '../question/actions'
import { setNotification } from '../notifications/actions'
import { deleteQuiz } from '../quizSearch/actions'
import { setQuiz, updateQuizStatus } from './actions'
import { hideModal, showModal2 } from '../modal/actions'
import { del } from '../axiosRequests/requests'
import history from '../history'

import '../stylesheets/quiz.css'

class Quiz extends React.Component {

  handleDeleteQuestion = () => {
    const { currentQuestion, userData, setNotification, deleteQuestion, hideModal } = this.props;
    const config = {
      data: [currentQuestion]
    }
    del("question/delete", config, userData.jwt).then((response) => {
      hideModal()
      deleteQuestion(currentQuestion);
      setNotification("Question deleted", "success", true);
    }).catch((error) => {
      console.log(error.response);
      hideModal()
      setNotification("Error - Unable to delete this question", "error", true)
    });
  }

  handleDeleteQuiz = () => {
    const { quiz, userData, deleteQuiz, setNotification } = this.props;
    const config = {
      data: quiz
    }
    del("quiz/delete", config, userData.jwt).then((response) => {
      hideModal()
      deleteQuiz(quiz)
      history.push("/quizSearch")
      setNotification("Quiz deleted", "success", true)
    }).catch((error) => {
      console.log(error.response)
      hideModal()
      setNotification("Error - Unable to delete this quiz", "error", true)
    })
  }

  renderDetails = () => {
    const { name, description } = this.props.quiz
    return(
      <React.Fragment>
        <div id="quizTitle">{name}</div>
        <div className="quizStatus">Status: {this.renderStatus()}</div>
        <div>Description:</div>
        <div>{description}</div>
      </React.Fragment>
    );
  }

  renderUpdateForm = () => {
    const { name } = this.props.quiz
    return(
      <React.Fragment>
        <div id="quizTitle">{`Edit the ${name} quiz`}</div>
        <div className="quizStatus">Status: {this.renderStatus()}</div>
        <UpdateQuizForm />
      </React.Fragment>
    );
  }

  renderDetailsOrUpdate = () => {
    return history.location.pathname === "/editQuiz" ? this.renderUpdateForm() : this.renderDetails()
  }

  renderQuestions = () => {
    const { questions } = this.props;
    return questions.length === 0 ? null : <Questions questions={questions}/>
  }

  renderAddButton = () => {
    const { permission } = this.props.userData
    return permission === "ADMIN" ? <Link to="/newQuestion" className="add addButton"><i className="fas fa-plus-circle green"></i> Add a question</Link> : null
  }

  triggerModal = () => {
    this.props.showModal2()
  }

  renderDeleteButton = () => {
    return <button onClick={this.triggerModal} className="delete"><i class="fas fa-trash-alt"></i> Delete</button>
  }

  renderStatusButton = () => {
    const { quiz } = this.props;
    return( <button className={quiz.status === "DRAFT" ? "save" : "warningButton"} onClick={this.updateStatus}>
              { quiz.status === "DRAFT" ? <i className="far fa-check-circle"></i> : <i className="fas fa-pencil-ruler"></i>}
              { quiz.status === "DRAFT" ? " Mark as Ready" : " Revert to draft" }
            </button>
    );
  }

  updateStatus = () => {
    const { quiz, updateQuizStatus, userData } = this.props;
    const data = {
      id: quiz.id,
      name: quiz.name,
      description: quiz.description,
      category: quiz.category,
      status: quiz.status === "DRAFT" ? "READY" : "DRAFT"
    }
    updateQuizStatus(data, userData.jwt)
  }

  renderStatus = () => {
    const { quiz } = this.props;
    return quiz.status === "DRAFT" ? <span className="draftFlag"><i className="fas fa-pencil-ruler"></i> Currenty in Draft</span> : <span className="readyFlag"><i className="far fa-check-circle"></i> Ready for use</span>
  }

  renderOptions = () => {
    const { permission } = this.props.userData
    return permission === "ADMIN" ? <div>{this.renderDeleteButton()}{this.renderStatusButton()}</div> : null
  }



  render(){
    const { questions, currentQuestion, hideModal, modalState } = this.props
    return(
      <div id="quiz">
        <Modal
          show={modalState.showModal}
          title={"Question"}
          message={`You are about to delete question ${currentQuestion.questionNumber}, this will also delete any associated answers for this question.`}
          onDelete={this.handleDeleteQuestion}
          onCancel={hideModal}
        />
        <Modal
          show={modalState.showModal2}
          title={"Quiz"}
          message={"You are about to delete a quiz which will also delete any questions and answers associated with this"}
          onDelete={this.handleDeleteQuiz}
          onCancel={hideModal}
        />
        <Notification />
        {this.renderDetailsOrUpdate()}
        {this.renderOptions()}
        <div className="headerContainer">
          <div id="questionsTitle">Questions</div>
          {this.renderAddButton()}
        </div>
        <div id="questionHeadings">
          <div id="questionNumberHeader">Answer</div>
          <div id="questionDescriptionHeader">Description</div>
          <div id="questionOptionsHeader"></div>
        </div>
        <Questions questions={questions}/>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    userData: state.userData,
    quiz: state.quiz,
    questions: state.questions,
    currentQuestion: state.currentQuestion,
    modalState: state.showModal
  }
}

export default connect(mapStateToProps,
  { getQuestions,
    setQuiz,
    setNotification,
    hideModal,
    deleteQuestion,
    deleteQuiz,
    showModal2,
    updateQuizStatus
  })(Quiz)
