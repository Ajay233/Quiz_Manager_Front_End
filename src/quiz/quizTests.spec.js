import React from 'react'
import ReactDOM from 'react-dom'
import { Quiz, mapStateToProps } from './quiz'
import { Router } from 'react-router-dom'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import mockAxios from 'jest-mock-axios'
import { render, fireEvent, cleanup } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { getQuestions, deleteQuestion, setCurrentQuestion } from '../question/actions'
import { setNotification } from '../notifications/actions'
import { deleteQuiz } from '../quizSearch/actions'
import { getAnswers } from '../answer/actions'
import { setQuiz, updateQuizStatus } from './actions'
import { hideModal, showModal2, showModal } from '../modal/actions'
import { sessionExpired } from '../utils/session'
import history from '../history'

const mockStore = configureStore({})

jest.mock('../question/actions')
jest.mock('../notifications/actions')
jest.mock('../quizSearch/actions')
jest.mock('../answer/actions')
jest.mock('./actions')
jest.mock('../modal/actions')
jest.mock('../utils/session')

describe("mapStateToProps", () => {
  it("should map state to props", () => {
    const appState = {
      userData: { id: 1 },
      quiz: { id: 1, name: "test", description: "test" },
      modalState: { showModal: false, showModal2: false, showModal3: false },
      questions: [
        { id: 1, questionNumber: 1, description: "test" },
        { id: 2, questionNumber: 2, description: "test2" }
      ],
      currentQuestion: { id: 1, questionNumber: 1, description: "test" },
      lists: ["Cat1", "Cat2", "Cat3"]
    }

    const componentState = mapStateToProps(appState)
    expect(componentState).toEqual(appState)
  })
})

describe("Quiz", () => {

  let store;
  let allModalsClosed;
  let currentQuestion;
  let quiz;
  let userData;
  let questions;


  beforeEach(() => {
    store = mockStore({
      lists: { categories: ["Cat1", "Cat2", "Cat3"] },
      notificationData: { message: "", type: "", show: false, timed: true},
      quiz: { id: 1, name: "test", description: "test" }
    })

    allModalsClosed = { showModal: false, showModal2: false, showModal3: false }
    quiz = { id: 1, name: "test", description: "test", category: "test", status: "READY" },
    currentQuestion =  { id: 1, questionNumber: 1, description: "test" }
    userData = { id: 1 }
    questions = [
      { id: 1, questionNumber: 1, description: "test" },
      { id: 2, questionNumber: 2, description: "test2" }
    ]

    ReactDOM.createPortal = jest.fn((element, node) => {
      return element
    })
  })

  it("should render the quiz details, 'READY' status and questions for '/viewQuiz' url", () => {
    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <Quiz
            userData={userData}
            modalState={allModalsClosed}
            currentQuestion={currentQuestion}
            quiz={quiz}
            questions={questions}
          />
        </Router>
      </Provider>
    )

    expect(component).toMatchSnapshot()
  })

  it("should render the quiz form, 'DRAFT' status and questions for '/editQuiz' url", () => {
    userData = { id: 1, permission: 'ADMIN' }
    let draftQuiz = { id: 1, name: "test", description: "test", category: "test", status: "READY" }
    history.push('/editQuiz')
    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <Quiz
            userData={userData}
            modalState={allModalsClosed}
            currentQuestion={currentQuestion}
            quiz={draftQuiz}
            questions={questions}
          />
        </Router>
      </Provider>
    )

    expect(component).toMatchSnapshot()
  })

  describe("Delete quiz button", () => {

    it("should call showModal2", () => {
      cleanup()
      userData = { id: 1, permission: 'ADMIN' }
      history.push('/editQuiz')
      const component = render(
        <Provider store={store}>
          <Router history={history}>
            <Quiz
              userData={userData}
              modalState={allModalsClosed}
              currentQuestion={currentQuestion}
              quiz={quiz}
              questions={questions}
              showModal2={showModal2}
            />
          </Router>
        </Provider>
      )

      fireEvent.click(component.getByTestId("delete-quiz-button"))
      expect(showModal2).toHaveBeenCalledTimes(1)
    })
  })

  describe("handleDeleteQuestion", () => {

    let component;
    let stateToRenderModal1

    beforeEach(() => {
      userData = { id: 1, permission: 'ADMIN' }
      stateToRenderModal1 = { showModal: true, showModal2: false, showModal3: false }
      history.push('/editQuiz')

      component = render(
        <Provider store={store}>
          <Router history={history}>
            <Quiz
              userData={userData}
              modalState={stateToRenderModal1}
              currentQuestion={currentQuestion}
              quiz={quiz}
              questions={questions}
              hideModal={hideModal}
              deleteQuestion={deleteQuestion}
              setNotification={setNotification}
            />
          </Router>
        </Provider>
      )
    })

    afterEach(() => {
      hideModal.mockClear()
      setNotification.mockClear()
      sessionExpired.mockClear()
      mockAxios.reset()
    })

    it("should call hideModal and deleteQuestion and then setNotification", () => {
      const requestResponse = {
        data: "DELETED"
      }
      fireEvent.click(component.getByTestId("modal-delete-button"))
      mockAxios.mockResponse(requestResponse)
      expect(hideModal).toHaveBeenCalledTimes(1)
      expect(deleteQuestion).toHaveBeenCalledTimes(1)
      expect(setNotification).toHaveBeenCalledTimes(1)
      expect(setNotification).toHaveBeenCalledWith("Question deleted", "success", true)
    })

    it("should call sessionExpired if the error status is 403", () => {
      const errorResponse = {
        response: {
          status: 403
        }
      }
      fireEvent.click(component.getByTestId("modal-delete-button"))
      mockAxios.mockError(errorResponse)
      expect(sessionExpired).toHaveBeenCalledTimes(1)
    })

    it("should call hideModal and setNotification for any other error", () => {
      const errorResponse = {
        response: {
          status: 404
        }
      }
      fireEvent.click(component.getByTestId("modal-delete-button"))
      mockAxios.mockError(errorResponse)
      expect(hideModal).toHaveBeenCalledTimes(1)
      expect(setNotification).toHaveBeenCalledWith("Error - Unable to delete this question", "error", true)
    })
  })

  describe("handleDeleteQuiz", () => {

    let component;
    let stateToRenderModal2

    beforeEach(() => {
      userData = { id: 1, permission: 'ADMIN' }
      stateToRenderModal2 = { showModal: false, showModal2: true, showModal3: false }
      history.push('/editQuiz')

      component = render(
        <Provider store={store}>
          <Router history={history}>
            <Quiz
              userData={userData}
              modalState={stateToRenderModal2}
              currentQuestion={currentQuestion}
              quiz={quiz}
              questions={questions}
              hideModal={hideModal}
              deleteQuiz={deleteQuiz}
              setNotification={setNotification}
            />
          </Router>
        </Provider>
      )
    })

    afterEach(() => {
      hideModal.mockClear()
      setNotification.mockClear()
      sessionExpired.mockClear()
      mockAxios.reset()
    })

    it("should call hideModal and deleteQuiz before redirectng to 'quizSearch' and calling setNotification", () => {
      const requestResponse = {
        data: "DELETED"
      }
      history.push = jest.fn()
      fireEvent.click(component.getByTestId("modal-delete-button"))
      mockAxios.mockResponse(requestResponse)
      expect(hideModal).toHaveBeenCalledTimes(1)
      expect(deleteQuiz).toHaveBeenCalledTimes(1)
      expect(history.push).toHaveBeenCalledTimes(1)
      expect(setNotification).toHaveBeenCalledTimes(1)
      expect(setNotification).toHaveBeenCalledWith("Quiz deleted", "success", true)
    })

    it("should call sessionExpired if the error status is 403", () => {
      const errorResponse = {
        response: {
          status: 403
        }
      }
      fireEvent.click(component.getByTestId("modal-delete-button"))
      mockAxios.mockError(errorResponse)
      expect(sessionExpired).toHaveBeenCalledTimes(1)
    })

    it("should call hideModal and setNotification for any other error", () => {
      const errorResponse = {
        response: {
          status: 404
        }
      }
      fireEvent.click(component.getByTestId("modal-delete-button"))
      mockAxios.mockError(errorResponse)
      expect(hideModal).toHaveBeenCalledTimes(1)
      expect(setNotification).toHaveBeenCalledWith("Error - Unable to delete this quiz", "error", true)
    })
  })
})
