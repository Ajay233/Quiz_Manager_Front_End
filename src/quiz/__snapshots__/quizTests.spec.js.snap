// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz should render the quiz details, 'READY' status and questions for '/viewQuiz' url 1`] = `
<div
  id="quiz"
>
  <div
    className="spacing"
    id="notification"
  />
  <div
    id="quizTitle"
  >
    test
  </div>
  <div
    className="quizStatus"
  >
    Status: 
    <span
      className="readyFlag"
    >
      <i
        className="far fa-check-circle"
      />
       Ready for use
    </span>
  </div>
  <div>
    Description:
  </div>
  <div>
    test
  </div>
  <div
    className="headerContainer"
  >
    <div
      id="questionsTitle"
    >
      Questions
    </div>
  </div>
  <div
    id="questionHeadings"
  >
    <div
      id="questionNumberHeader"
    >
      Answer
    </div>
    <div
      id="questionDescriptionHeader"
    >
      Description
    </div>
    <div
      id="questionOptionsHeader"
    />
  </div>
  <div>
    <div>
      <div
        className="questionContainer"
      >
        <div
          className="questionNumber"
        >
          1
        </div>
        <div
          className="questionDescription"
        >
          test
        </div>
        <div
          className="options"
        />
      </div>
    </div>
    <div>
      <div
        className="questionContainer"
      >
        <div
          className="questionNumber"
        >
          2
        </div>
        <div
          className="questionDescription"
        >
          test2
        </div>
        <div
          className="options"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Quiz should render the quiz form, 'DRAFT' status and questions for '/editQuiz' url 1`] = `
<div
  id="quiz"
>
  <div
    className="spacing"
    id="notification"
  />
  <div
    id="quizTitle"
  >
    Edit the test quiz
  </div>
  <div
    className="quizStatus"
  >
    Status: 
    <span
      className="readyFlag"
    >
      <i
        className="far fa-check-circle"
      />
       Ready for use
    </span>
  </div>
  <div>
    <form
      onSubmit={[Function]}
    >
      <div>
        <label>
          Quiz name:
        </label>
        <input
          className="inputBox"
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          placeholder="test"
          value=""
        />
      </div>
      <div>
        <label>
          Quiz description:
        </label>
        <input
          className="inputBox"
          name="description"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          placeholder="test"
          value=""
        />
      </div>
      <div>
        <label>
          Quiz category:
        </label>
        <select
          className="inputBox"
          name="category"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          value=""
        >
          <option
            value="Cat1"
          >
            Cat1
          </option>
          <option
            value="Cat2"
          >
            Cat2
          </option>
          <option
            value="Cat3"
          >
            Cat3
          </option>
        </select>
      </div>
      <div>
        <button
          className="submit"
        >
          Save
        </button>
        <a
          className="cancel"
          href="/quizSearch"
          onClick={[Function]}
        >
          Cancel
        </a>
      </div>
    </form>
  </div>
  <div>
    <button
      className="delete"
      data-testid="delete-quiz-button"
      onClick={[Function]}
    >
      <i
        className="fas fa-trash-alt"
      />
       Delete
    </button>
    <button
      className="warningButton"
      data-testid="updateStatus-button"
      onClick={[Function]}
    >
      <i
        className="fas fa-pencil-ruler"
      />
       Revert to draft
    </button>
  </div>
  <div
    className="headerContainer"
  >
    <div
      id="questionsTitle"
    >
      Questions
    </div>
    <a
      className="add addButton"
      href="/newQuestion"
      onClick={[Function]}
    >
      <i
        className="fas fa-plus-circle green"
      />
       Add a question
    </a>
  </div>
  <div
    id="questionHeadings"
  >
    <div
      id="questionNumberHeader"
    >
      Answer
    </div>
    <div
      id="questionDescriptionHeader"
    >
      Description
    </div>
    <div
      id="questionOptionsHeader"
    />
  </div>
  <div>
    <div>
      <div
        className="questionContainer"
      >
        <div
          className="questionNumber"
        >
          1
        </div>
        <div
          className="questionDescription"
        >
          test
        </div>
        <div
          className="options"
        >
          <a
            className="edit"
            href="/editQuestion"
            onClick={[Function]}
          >
            <i
              className="fas fa-edit blue"
            />
             Edit
          </a>
          <a
            className="deleteOption"
            href="/editQuiz"
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt red"
            />
             Delete
          </a>
        </div>
      </div>
    </div>
    <div>
      <div
        className="questionContainer"
      >
        <div
          className="questionNumber"
        >
          2
        </div>
        <div
          className="questionDescription"
        >
          test2
        </div>
        <div
          className="options"
        >
          <a
            className="edit"
            href="/editQuestion"
            onClick={[Function]}
          >
            <i
              className="fas fa-edit blue"
            />
             Edit
          </a>
          <a
            className="deleteOption"
            href="/editQuiz"
            onClick={[Function]}
          >
            <i
              className="fas fa-trash-alt red"
            />
             Delete
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
